### PR TITLE
other mangleing

### PR DESCRIPTION
**Pull request by ZuseZ4**
Just to not forget about it.

### CODE REVIEW

File: enzyme/Enzyme/AdjointGenerator.h [hunk 1]
```diff
@@ -8944,6 +8944,8 @@ class AdjointGenerator
       // if constant instruction and readonly (thus must be pointer return)
       // and shadow return recomputable from shadow arguments.
       if (funcName == "__dynamic_cast" ||
+          funcName == "_ZSt18_Rb_tree_decrementPKSt18_Rb_tree_node_base" ||
+          funcName == "_ZSt18_Rb_tree_incrementPKSt18_Rb_tree_node_base" ||
           funcName == "_ZSt18_Rb_tree_decrementPSt18_Rb_tree_node_base" ||
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:8949]: Can you also update type analysis with this?

**Comment by wsmoses** [comment]
Would you be able to add a test using these?

**Comment by ZuseZ4** [comment]
> Would you be able to add a test using these?

Can I just copy and adjust the ReverseMode/doubleerase.ll test which we have?
The rb_tree impl isn't directly available from C++ and I can't use my minimized OpenLB code due to a ptr that isn't arg issue.
Constructing it by hand on llvm-ir seems a bit cumbersome.

**Comment by wsmoses** [comment]
If you're going to add an ll test you should just construct one that just calls these functions.

I'd, however, also recommend making a minimal std::map test that does the various operations you care about (and in doing so calls these functions) to ensure that the behavior is maintained (and we can then see any mangling differences from CI/tests).

**Comment by ZuseZ4** [comment]
I just added a small handwritten testcase, however I hit the gutils->isConstantInst(orig) assertion:
```.sh
  1   opt: /home/drehwald/prog/Enzyme/enzyme/Enzyme/AdjointGenerator.h:9107: void AdjointGenerator<const AugmentedReturn *>::visitCallInst(llvm::CallInst &) [AugmentedRe      turnType = const AugmentedReturn *]: Assertion `gutils->isConstantInstruction(orig)' failed.
    1 PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
    2 Stack dump:
    3 0.      Program arguments: opt rb_tree.ll -O3 --enable-new-pm=0 -load=/home/drehwald/prog/Enzyme/enzyme/build/Enzyme/LLVMEnzyme-14.so -enzyme -enzyme-strict-aliasi      ng=0 -enzyme-loose-types=1 -S -o out.ll
    4 1.      Running pass 'Enzyme Pass' on module 'rb_tree.ll'.
    5  #0 0x000000000297f753 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/drehwald/prog/llvm-14/build/bin/opt+0x297f753)
    6  #1 0x000000000297d52e llvm::sys::RunSignalHandlers() (/home/drehwald/prog/llvm-14/build/bin/opt+0x297d52e)
    7  #2 0x000000000297fc1a SignalHandler(int) Signals.cpp:0:0
    8  #3 0x00007fc38f77d520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
    9  #4 0x00007fc38f7d1a7c __pthread_kill_implementation ./nptl/./nptl/pthread_kill.c:44:76
   10  #5 0x00007fc38f7d1a7c __pthread_kill_internal ./nptl/./nptl/pthread_kill.c:78:10
   11  #6 0x00007fc38f7d1a7c pthread_kill ./nptl/./nptl/pthread_kill.c:89:10
   12  #7 0x00007fc38f77d476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
   13  #8 0x00007fc38f7637f3 abort ./stdlib/./stdlib/abort.c:81:7
   14  #9 0x00007fc38f76371b _nl_load_domain ./intl/./intl/loadmsgcat.c:1177:9
   15 #10 0x00007fc38f774e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
   16 #11 0x00007fc38f4bb932 AdjointGenerator<AugmentedReturn const*>::visitCallInst(llvm::CallInst&) /home/drehwald/prog/Enzyme/enzyme/Enzyme/AdjointGenerator.h:0:9
   17 #12 0x00007fc38f4a1a7b llvm::InstVisitor<AdjointGenerator<AugmentedReturn const*>, void>::delegateCallInst(llvm::CallInst&) /home/drehwald/prog/llvm-14/llvm/includ      e/llvm/IR/InstVisitor.h:303:3
   18 #13 0x00007fc38f48c3ad llvm::InstVisitor<AdjointGenerator<AugmentedReturn const*>, void>::visitCall(llvm::CallInst&) /home/drehwald/prog/llvm-14/llvm/include/llvm/      IR/Instruction.def:209:1
   19 #14 0x00007fc38f48b86a llvm::InstVisitor<AdjointGenerator<AugmentedReturn const*>, void>::visit(llvm::Instruction&) /home/drehwald/prog/llvm-14/llvm/include/llvm/I      R/Instruction.def:209:1
   20 #15 0x00007fc38f3937bd llvm::InstVisitor<AdjointGenerator<AugmentedReturn const*>, void>::visit(llvm::Instruction*) /home/drehwald/prog/llvm-14/llvm/include/llvm/I      R/InstVisitor.h:112:34
   21 #16 0x00007fc38f3767ac EnzymeLogic::CreatePrimalAndGradient(ReverseCacheKey const&&, TypeAnalysis&, AugmentedReturn const*, bool) /home/drehwald/prog/Enzyme/enzyme      /Enzyme/EnzymeLogic.cpp:3862:7
   22 #17 0x00007fc38f32cac5 (anonymous namespace)::Enzyme::HandleAutoDiff(llvm::CallInst*, llvm::TargetLibraryInfo&, DerivativeMode, bool) /home/drehwald/prog/Enzyme/en      zyme/Enzyme/Enzyme.cpp:1580:15
   23 #18 0x00007fc38f328b82 (anonymous namespace)::Enzyme::lowerEnzymeCalls(llvm::Function&, bool&, std::set<llvm::Function*, std::less<llvm::Function*>, std::allocator      <llvm::Function*> >&) /home/drehwald/prog/Enzyme/enzyme/Enzyme/Enzyme.cpp:2280:21
   24 #19 0x00007fc38f321e60 (anonymous namespace)::Enzyme::runOnModule(llvm::Module&) /home/drehwald/prog/Enzyme/enzyme/Enzyme/Enzyme.cpp:2477:18
   25 #20 0x00000000020aa87c llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/drehwald/prog/llvm-14/build/bin/opt+0x20aa87c)

```
It looks to me like another rb_tree related enzyme bug, or did I just miss something in my testcase?

**Comment by wsmoses** [comment]
I think you also need to add your manglings to ActivityAnalysis here: https://github.com/EnzymeAD/Enzyme/blob/563c88fbe725f2a29fca10bf8bbd4bcb490fc3e5/enzyme/Enzyme/ActivityAnalysis.cpp#L189

**Review by wsmoses — APPROVED** [review]

**Pull request merged**